src/scripts/clipperUI/ratingsHelper.ts

```diff
@@ -0,0 +1,294 @@
+	 *      a) The major version last seen by the user is greater than the major version at the time of the last bad rating, OR
+	 *      b) The major version remained the same, while the minor version last seen by the user is greater than
+	 *          the minor version at the time of the last bad rating
+	 *
+	 * Public for testing
+	 */
+	public static badRatingVersionDelayIsOver(badRatingVersionAsStr: string, lastSeenVersionAsStr: string): boolean {
+		if (Utils.isNullOrUndefined(badRatingVersionAsStr)) {
+			// value has never been set, no bad rating given
+			return true;
+		}
+
+		let badRatingVersion: Version;
+		let lastSeenVersion: Version;
+		try {
+			badRatingVersion = new Version(badRatingVersionAsStr);
+			lastSeenVersion = new Version(lastSeenVersionAsStr);
+		} catch (e) {
+			return false;
+		}
+
+		return lastSeenVersion.isGreaterThan(badRatingVersion, true /* ignorePatchUpdate */);
+	}
+
+	/**
+	 * Returns true if ALL of the below applies:
+	 *   * Number of successful clips >= {Constants.Settings.minClipSuccessForRatingsPrompt}
+	 *   * Number of successful clips <= {Constants.Settings.maxClipSuccessForRatingsPrompt}
+	 *
+	 * Public for testing
+	 */
+	public static clipSuccessDelayIsOver(numClips: number): boolean {
+		// TODO MVP+? when # of successful clips > nMax, collapse panel into a Rate Us hyperlink in the footer that is always available
+
+		if (isNaN(numClips)) {
+			return false;
+		}
+
+		return numClips >= Constants.Settings.minClipSuccessForRatingsPrompt &&
+			numClips <= Constants.Settings.maxClipSuccessForRatingsPrompt;
+	}
+
+	/**
+	 * Implementation of the logic described in the setShowRatingsPromptState(...) description
+	 */
+	private static shouldShowRatingsPromptInternal(clipperState: ClipperState, event: Log.Event.PromiseEvent, logEventInfo: RatingsLoggingInfo): boolean {
+		if (Utils.isNullOrUndefined(clipperState)) {
+			event.setStatus(Log.Status.Failed);
+			event.setFailureInfo({ error: "Clipper state is null or undefined" });
+			return false;
+		}
+
+		if (!Utils.isNullOrUndefined(clipperState.showRatingsPrompt.get())) {
+			// return cached value in clipper state since it already exists
+			logEventInfo.usedCachedValue = true;
+			return clipperState.showRatingsPrompt.get();
+		}
+
+		let ratingsPromptEnabled: boolean = RatingsHelper.ratingsPromptEnabledForClient(clipperState.clientInfo.clipperType);
+		logEventInfo.ratingsPromptEnabledForClient = ratingsPromptEnabled;
+		if (!ratingsPromptEnabled) {
+			return false;
+		}
+
+		let doNotPromptRatingsStr: string = Clipper.Storage.getCachedValue(Constants.StorageKeys.doNotPromptRatings);
+		let lastBadRatingDateAsStr: string = Clipper.Storage.getCachedValue(Constants.StorageKeys.lastBadRatingDate);
+		let lastBadRatingVersion: string = Clipper.Storage.getCachedValue(Constants.StorageKeys.lastBadRatingVersion);
+		let lastSeenVersion: string = Clipper.Storage.getCachedValue(Constants.StorageKeys.lastSeenVersion);
```

Comment:
@VishaalK I definitely get your concern here - `getCachedValue` is easy to accidentally misuse.

@VishaalK @leakymemory @mttwc  I assume we're okay with using this paradigm for this PR, even if we need a better permanent solution... so how should we facilitate discussing the future

Looks like `async/await` didn't make it into ES7 (ES2016), and only ["has a chance" of making ES8 (ES2017)](http://www.2ality.com/2016/02/ecmascript-2017.html)... so do we want to polyfill it? I can open a task for investigation if we all agree.

And what happened with the first hard look at `async/await`?